ratis-test/src/test/java/org/apache/ratis/datastream/TestDataStreamBase.java

```diff
@@ -51,7 +54,22 @@ static byte pos2byte(int pos) {
     return (byte) ('A' + pos%MODULUS);
   }
 
-  static class SingleDataStreamStateMachine extends BaseStateMachine {
+  static class MultiDataStreamStateMachine extends BaseStateMachine {
+    final ConcurrentMap<Long, SingleDataStream> streams = new ConcurrentHashMap<>();
+
+    @Override
+    public CompletableFuture<DataStream> stream(RaftClientRequest request) {
+      final SingleDataStream s = new SingleDataStream();
+      streams.put(request.getCallId(), s);
```

Comment:
I am thinking we should use `stream id` than `call id` to create `SingleDataStream` for each id?

The call id is created at https://github.com/apache/incubator-ratis/blob/865a6cab64313e6702d557c385ecd328abd43692/ratis-client/src/main/java/org/apache/ratis/client/impl/OrderedAsync.java#L165, which is a monotonic increasing id.

Reply:
Ok after dig into the implementation, currently `stream(RaftClientRequest request)` is called for header requests. Thus `call id` will be unique for each stream as long as each stream only has one header and for that header, `stream(RaftClientRequest request)` is only called once.

This is just a test code so I guess this does not matter much.

Reply:
Thanks for checking it.  I will see if we could make the streamId the same as the callId.